File: compress.py
#!/usr/bin/env python3

class Khabib:


    """
    This code consists of method pairs, compress and decompress, that work together to compress and decompress a string message based on a given code_book. The compress function , while the decompress function reverses this process to reconstruct the original string.

    This class is named in honor of Khabib Nurmagomedov, the undefeated UFC lightweight champion, who is known for his ability to compress his opponents into submission.
    ```
    """


def substring_compress_aggressive(code_book, message):
    """replaces substrings from message that match a substring in code_book with a compressed representation of the substring in the form <index,length> where index is the index of the substring in code_book and length is the length of the substring in code_book. The compressor will aggressively replace any substring that is found in the code_book, even if the compressed representation is longer than the original substring -- this should make the result more compressible with a general compression algorithm such as Deflate or LZMA."""
    compressed_message = ''
    current = ''
    
    for c in message:
        current += c

        if current not in code_book:
            compressed = f'<{code_book.index(current[:-1])},{len(current) - 1}>'
            compressed_message += compressed
            if c == '<':
                compressed_message += '<'
            compressed_message += c
            current = ''

    if len(current) > 0:
        compressed = f'<{code_book.index(current)},{len(current)}>'
        compressed_message += compressed

    return compressed_message


def substring_compress(code_book, message):
    return substring_compress_conservative(code_book, message)


def substring_compress_conservative(code_book, message):
    """replaces substrings from message that match a substring in code_book with a compressed representation of the substring in the form <index,length> where index is the index of the substring in code_book and length is the length of the substring in code_book. If the compressed representation is longer than the original substring, the original substring is returned."""

    compressed_message = ''
    current = ''

    for c in message:
        current += c

        if current not in code_book:
            compressed = f"<{code_book.index(current[:-1])},{len(current) - 1}>"
            if len(current) <= len(compressed):
                compressed_message += current.replace('<', '<<')
            else:
                compressed_message += compressed
                compressed_message += c.replace('<', '<<')
            current = ''

    if len(current) > 0:
        compressed = f"<{code_book.index(current)},{len(current)}>"
        if len(current) <= len(compressed):
            compressed_message += current.replace('<', '<<')
        else:
            compressed_message += compressed

    return compressed_message


def substring_decompress(code_book, compressed_message):
    decompressed_message = ''
    i = 0

    while i < len(compressed_message):
        c = compressed_message[i]

        if c == '<':
            if i + 1 < len(compressed_message) and compressed_message[i + 1] == '<':
                decompressed_message += '<'
                i += 2
            else:
                j = compressed_message.index(',', i + 1)
                start = int(compressed_message[i + 1:j])
                i = j + 1
                j = compressed_message.index('>', i)
                length = int(compressed_message[i:j])
                i = j + 1
                decompressed_message += code_book[start:start + length]
        else:
            decompressed_message += c
            i += 1

    return decompressed_message


def test_compression_decompression(code_book, message):
    compressed_message = substring_compress(code_book, message)
    decompressed_message = substring_decompress(code_book, compressed_message)

    print(f"Original message: {message}")
    print(f"Compressed message: {compressed_message}")
    print(f"Decompressed message: {decompressed_message}")

    if decompressed_message == message:
        # print("Test passed: substring_decompress(code_book, substring_compress(code_book, message)) == message")
        return True
    else:
        raise ("Test failed: substring_decompress(code_book, substring_compress(code_book, message)) != message")


def run_tests_aggressive():
    global substring_compress
    substring_compress = substring_compress_aggressive # XXX This is a hack to make the function available to the test function
    run_tests()


def run_tests():
    print("Running tests...")

    # Preliminary tests
    
    test_compression_decompression(
        'abcdefghijklmnopqrstuvwxyz', 'hello world! this is a test.')
    test_compression_decompression(
        'abcdefghijklmnopqrstuvwxyz', 'hello world! this is a test. hello world! this is a test.')
    test_compression_decompression(
        'one two three four five six seven eight nine ten', 'one two three four five six seven eight nine ten')
    test_compression_decompression(
        '/path/to file, with spaces', '/path/to file, with spaces')

    test_compression_decompression('["root/a 1.txt(abcd) 2.txt(efgh)","root/c 3.txt(abcd)","root/c/d 4.txt(efgh)","root 4.txt(efgh)"]',
                                   '[["root/a/2.txt","root/c/d/4.txt","root/4.txt"],["root/a/1.txt","root/c/3.txt"]]')

    # Test case 1
    system1 = 'abcdefghijklmnopqrstuvwxyz,foo,bar'
    target1 = 'hello world! foo,bar,foo,bar.'
    test_compression_decompression(system1, target1)

    print('------------------------------------------')

    # Test case 2
    system2 = '0123456789,one,two,three'
    target2 = 'one plus two equals three: 1 + 2 = 3'
    test_compression_decompression(system2, target2)

    print('------------------------------------------')

    # Test case 3
    system3 = 'aeiou,compression,decompression'
    target3 = 'compression and decompression'
    test_compression_decompression(system3, target3)

    # Test case 4
    code_book4 = 'abcdefghijklmnopqrstuvwxyz,The quick brown fox jumps over the lazy dog'
    message4 = 'The quick brown fox jumps over the lazy dog twice!'
    test_compression_decompression(code_book4, message4)

    print('------------------------------------------')

    # Test case 5
    code_book5 = '0123456789,Once upon a time in a land far far away'
    message5 = 'Once upon a time in a land far far away, there was a kingdom.'
    test_compression_decompression(code_book5, message5)

    print('------------------------------------------')

    # Test case 6
    code_book6 = 'aeiou,To be or not to be that is the question'
    message6 = 'To be or not to be that is the question: Shakespeare'
    test_compression_decompression(code_book6, message6)

    # Test case 7
    code_book7 = 'abcdefghijklmnopqrstuvwxyz,Once upon a time'
    message7 = 'Once upon a time, Alice found a message: <<Hello, World!>>'
    test_compression_decompression(code_book7, message7)

    print('------------------------------------------')


if __name__ == '__main__':
    print("Running tests...")
    print("Conservative tests:")
    run_tests()

    print("====================================================================================")
    print("Aggressive tests:")
    print("====================================================================================")

    run_tests_aggressive() # Call this last, as it changes the default compression function

File: compress.js
/** This code consists of two functions, compress and decompress, that work together to compress and decompress a string message based on a given codeBook. The compress function replaces substrings from message that match a substring in codeBook with a compressed representation, while the decompress function reverses this process to reconstruct the original string. */

function substringCompressAggressive(codeBook, message) {
    let compressedMessage = '';
    let current = '';

    for (const c of message) {
        current += c;

        if (!codeBook.includes(current)) {
            const compressed = `<${codeBook.indexOf(current.slice(0, -1))},${current.length - 1}>`;
            compressedMessage += compressed;
            compressedMessage += c.replace('<', '<<');
            current = '';
        }
    }

    if (current.length > 0) {
        const compressed = `<${codeBook.indexOf(current)},${current.length}>`;
        compressedMessage += compressed;
    }

    return compressedMessage;
}

function substringCompress(codeBook, message) {
    let compressedMessage = '';
    let current = '';

    for (const c of message) {
        current += c;

        if (!codeBook.includes(current)) {
            const compressed = `<${codeBook.indexOf(current.slice(0, -1))},${current.length - 1}>`;
            if (current.length <= compressed.length) {
                compressedMessage += current.replace('<', '<<');
            } else {
                compressedMessage += compressed;
                compressedMessage += c.replace('<', '<<');
            }
            current = '';
        }
    }

    if (current.length > 0) {
        const compressed = `<${codeBook.indexOf(current)},${current.length}>`;
        if (current.length <= compressed.length) {
            compressedMessage += current.replace('<', '<<');
        } else {
            compressedMessage += compressed;
        }
    }

    return compressedMessage;
}

function substringDecompress(codeBook, compressedMessage) {
    let decompressedMessage = '';
    let i = 0;

    while (i < compressedMessage.length) {
        const c = compressedMessage[i];

        if (c === '<') {
            if (i + 1 < compressedMessage.length && compressedMessage[i + 1] === '<') {
                decompressedMessage += '<';
                i += 2;
            } else {
                const j = compressedMessage.indexOf(',', i + 1);
                const start = parseInt(compressedMessage.slice(i + 1, j), 10);
                i = j + 1;
                const k = compressedMessage.indexOf('>', i);
                const length = parseInt(compressedMessage.slice(i, k), 10);
                i = k + 1;
                decompressedMessage += codeBook.slice(start, start + length);
            }
        } else {
            decompressedMessage += c;
            i++;
        }
    }

    return decompressedMessage;
}

function testCompressionDecompression(codeBook, message) {
    const compressedMessage = substringCompress(codeBook, message);
    const decompressedMessage = substringDecompress(codeBook, compressedMessage);

    console.log('Original message:', message);
    console.log('Compressed message:', compressedMessage);
    console.log('Decompressed message:', decompressedMessage);

    if (decompressedMessage === message) {
        console.log('Test passed: decompress(codeBook, compress(codeBook, message)) === message');
    } else {
        throw 'Test failed: decompress(codeBook, compress(codeBook, message)) !== message';
    }
}


function runTests() {
    // Test case 0
    const codeBook = 'abcdefghijklmnopqrstuvwxyz';
    const message = 'hello world! this is a test.';
    testCompressionDecompression(codeBook, message);

    console.log('------------------------------------------');

    // Test case 1
    const codeBook1 = 'abcdefghijklmnopqrstuvwxyz';
    const message1 = 'hello world! this is a test.';
    testCompressionDecompression(codeBook1, message1);

    console.log('------------------------------------------');

    // Test case 2
    const codeBook2 = '0123456789';
    const message2 = 'The total cost is $103.50 for 3 items.';
    testCompressionDecompression(codeBook2, message2);

    console.log('------------------------------------------');

    // Test case 3
    const codeBook3 = 'aeiou';
    const message3 = 'compression and decompression';
    testCompressionDecompression(codeBook3, message3);

    console.log('------------------------------------------');

    // Test case 4
    const codeBook4 = 'abcdefghijklmnopqrstuvwxyz,The quick brown fox jumps over the lazy dog';
    const message4 = 'The quick brown fox jumps over the lazy dog twice!';
    testCompressionDecompression(codeBook4, message4);

    console.log('------------------------------------------');

    // Test case 5
    const codeBook5 = '0123456789,Once upon a time in a land far far away';
    const message5 = 'Once upon a time in a land far far away, there was a kingdom.';
    testCompressionDecompression(codeBook5, message5);

    console.log('------------------------------------------');

    // Test case 6
    const codeBook6 = 'aeiou,To be or not to be that is the question';
    const message6 = 'To be or not to be that is the question: Shakespeare';
    testCompressionDecompression(codeBook6, message6);

    // Test case 7
    const codeBook7 = 'abcdefghijklmnopqrstuvwxyz,Once upon a time';
    const message7 = 'Once upon a time, Alice found a message: <<Hello, World!>>';
    testCompressionDecompression(codeBook7, message7);

    console.log('------------------------------------------');
}

function runTestsAggressive() {
    substringCompress = substringCompressAggressive; // XXX This is a hack to make the function available to the test function
    runTests();
}

runTests();
runTestsAggressive(); // run this last as it modifies the compress function
